serve/plugin.go

```diff
@@ -140,45 +136,12 @@ func (s *PluginServe) newCmdPluginServe() *cobra.Command {
 				logger = log.Output(zerolog.ConsoleWriter{Out: os.Stdout}).Level(zerologLevel)
 			}
 
-			if otelEndpoint != "" {
-				resource := newResource(s.plugin)
-				opts := []otlptracehttp.Option{
-					otlptracehttp.WithEndpoint(otelEndpoint),
-				}
-				if otelEndpointInsecure {
-					opts = append(opts, otlptracehttp.WithInsecure())
-				}
-				if len(otelEndpointHeaders) > 0 {
-					headers := make(map[string]string, len(otelEndpointHeaders))
-					for _, h := range otelEndpointHeaders {
-						parts := strings.SplitN(h, ":", 2)
-						if len(parts) != 2 {
-							return fmt.Errorf("invalid header %q", h)
-						}
-						headers[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
-					}
-					opts = append(opts, otlptracehttp.WithHeaders(headers))
-				}
-				if otelEndpointURLPath != "" {
-					opts = append(opts, otlptracehttp.WithURLPath(otelEndpointURLPath))
-				}
-
-				client := otlptracehttp.NewClient(opts...)
-				exp, err := otlptrace.New(cmd.Context(), client)
-				if err != nil {
-					return fmt.Errorf("creating OTLP trace exporter: %w", err)
-				}
-				tp := trace.NewTracerProvider(
-					trace.WithBatcher(exp),
-					trace.WithResource(resource),
-				)
-				defer func() {
-					if err := tp.Shutdown(context.Background()); err != nil {
-						logger.Error().Err(err).Msg("failed to shutdown OTLP trace exporter")
-					}
-				}()
-				otel.SetTracerProvider(tp)
+			shutdown, err := setupOtel(cmd.Context(), logger, s.plugin, otelEndpoint, otelEndpointInsecure, otelEndpointHeaders, otelEndpointURLPath)
+			if err != nil {
+				return fmt.Errorf("failed to setup OpenTelemetry: %w", err)
 			}
+			defer shutdown()
```

Comment:
I think if `err != nil` this will blow up at defer time (e.g. https://go.dev/play/p/dOvHV41W1CE)

Reply:
Shouldn't returning early when `err != nil` prevent us from reaching `defer shutdown()` ? 
https://github.com/cloudquery/plugin-sdk/pull/1751/files#diff-9d26691c28ec41ba1e4872ba4df8c1574c2ffc6582c6b82a65914f4e301a7517R141

Reply:
Ah, I didn't expect that a failure to setup otel would cause the serve plugin to fail. Is that ok?

Reply:
So we fail on setting OTEL when you explicitly pass the endpoint (same behavior as before) https://github.com/cloudquery/plugin-sdk/blob/e5e8e7ea650862e53214f381db6c22173fb04edb/serve/plugin.go#L144 and https://github.com/cloudquery/plugin-sdk/blob/e5e8e7ea650862e53214f381db6c22173fb04edb/serve/plugin.go#L169